pkg/controller/bootstrap/tokencleaner.go

```diff
@@ -59,57 +62,127 @@ type TokenCleaner struct {
+		queue:                workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "token_cleaner"),
 	}
+
 	if cl.CoreV1().RESTClient().GetRateLimiter() != nil {
 		if err := metrics.RegisterMetricAndTrackRateLimiterUsage("token_cleaner", cl.CoreV1().RESTClient().GetRateLimiter()); err != nil {
 			return nil, err
 		}
 	}
 
-	secretSelector := fields.SelectorFromSet(map[string]string{api.SecretTypeField: string(bootstrapapi.SecretTypeBootstrapToken)})
-	e.secrets, e.secretsController = cache.NewInformer(
-		&cache.ListWatch{
-			ListFunc: func(lo metav1.ListOptions) (runtime.Object, error) {
-				lo.FieldSelector = secretSelector.String()
-				return e.client.CoreV1().Secrets(e.tokenSecretNamespace).List(lo)
+	secrets.Informer().AddEventHandlerWithResyncPeriod(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1.Secret:
+					return t.Type == bootstrapapi.SecretTypeBootstrapToken && t.Namespace == e.tokenSecretNamespace
+				default:
+					utilruntime.HandleError(fmt.Errorf("object passed to %T that is not expected: %T", e, obj))
+					return false
+				}
 			},
-			WatchFunc: func(lo metav1.ListOptions) (watch.Interface, error) {
-				lo.FieldSelector = secretSelector.String()
-				return e.client.CoreV1().Secrets(e.tokenSecretNamespace).Watch(lo)
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc:    e.enqueueSecrets,
+				UpdateFunc: func(oldSecret, newSecret interface{}) { e.enqueueSecrets(newSecret) },
 			},
 		},
-		&v1.Secret{},
 		options.SecretResync,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    e.evalSecret,
-			UpdateFunc: func(oldSecret, newSecret interface{}) { e.evalSecret(newSecret) },
-		},
 	)
+
 	return e, nil
 }
 
 // Run runs controller loops and returns when they are done
 func (tc *TokenCleaner) Run(stopCh <-chan struct{}) {
-	go tc.secretsController.Run(stopCh)
-	go wait.Until(tc.evalSecrets, 10*time.Second, stopCh)
+	defer utilruntime.HandleCrash()
+
+	glog.Infof("Starting token cleaner controller")
+	defer glog.Infof("Shutting down token cleaner controller")
+
+	if !controller.WaitForCacheSync("token_cleaner", stopCh, tc.secretSynced) {
+		return
+	}
+
+	go wait.Until(tc.worker, 10*time.Second, stopCh)
+
 	<-stopCh
 }
 
-func (tc *TokenCleaner) evalSecrets() {
-	for _, obj := range tc.secrets.List() {
-		tc.evalSecret(obj)
+func (tc *TokenCleaner) enqueueSecrets(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %#v: %v", obj, err)
+		return
+	}
+	tc.queue.Add(key)
+}
+
+// worker runs a thread that dequeues CSRs, handles them, and marks them done.
+func (tc *TokenCleaner) worker() {
+	for tc.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (tc *TokenCleaner) processNextWorkItem() bool {
+	cKey, quit := tc.queue.Get()
```

Comment:
Just call it `key`? I'm assuming `cKey` came from certificate_controller.go?

Reply:
Yes, Updated